### PR TITLE
Optimize CSE + FeatureFlags option for run in background

### LINKED ISSUE #4126: Can we increase the timeout for docker pull?
**Is this a request for help?**:
Yes,
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE
---

**What version of acs-engine?**:
Version: v0.24.1
GitCommit: 1ff11132
GitTreeState: clean
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes.

**What happened**:
the docker pull timed out:

`+ retrycmd_if_failure 100 1 60 docker pull crproxy.trafficmanager.net:6000/google_containers/hyperkube-amd64:v1.8.4
+ retries=100
+ wait=1
+ timeout=60
+ shift
+ shift
+ shift
++ seq 1 100
+ for i in '$(seq 1 $retries)'
+ timeout 60 docker pull crproxy.trafficmanager.net:6000/google_containers/hyperkube-amd64:v1.8.4
v1.8.4: Pulling from google_containers/hyperkube-amd64
bfa86312bfc6: Pulling fs layer
deebd8771aa2: Pulling fs layer
102196268dd2: Pulling fs layer
f205dd477573: Pulling fs layer
4e69eb904993: Pulling fs layer
733382c27720: Pulling fs layer
f205dd477573: Waiting
4e69eb904993: Waiting
733382c27720: Waiting
deebd8771aa2: Verifying Checksum
deebd8771aa2: Download complete
f205dd477573: Verifying Checksum
f205dd477573: Download complete
4e69eb904993: Download complete
bfa86312bfc6: Verifying Checksum
bfa86312bfc6: Download complete
bfa86312bfc6: Pull complete
deebd8771aa2: Pull complete
+ '[' 124 -eq 0 ']'
+ sleep 1
+ for i in '$(seq 1 $retries)'
+ timeout 60 docker pull crproxy.trafficmanager.net:6000/google_containers/hyperkube-amd64:v1.8.4
v1.8.4: Pulling from google_containers/hyperkube-amd64
bfa86312bfc6: Pulling fs layer
deebd8771aa2: Pulling fs layer
102196268dd2: Pulling fs layer
f205dd477573: Pulling fs layer
4e69eb904993: Pulling fs layer
733382c27720: Pulling fs layer
f205dd477573: Waiting
4e69eb904993: Waiting
733382c27720: Waiting`

**What you expected to happen**:
I think we should give more time for the docker image to be pulled

**How to reproduce it** (as minimally and precisely as possible):
I am trying to deploy a cluster in China. I am not sure why it takes a lot of time. 

**Anything else we need to know**:
I sshd to the vm and pulled the image to the vm and the process continued. I think some of the delay created during the deployment of acs engine is caused by this timeout.

### PR DESCRIPTION
**Pull request by jackfrancis**
**What this PR does / why we need it**: Improves sanity enforcement in CSE:

- wait for files before `source`ing them
- increase retries/timeouts everywhere to accommodate transient cloud-ness and fulfill "eventual success" strategy
- move outbound access test to main CSE invocation, so that we can re-use main provision scripts in blocking (masters) and non-blocking (nodes) contexts
- add a bunch of resilience around GPU drivers installation/configuration

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #4126

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```
Optimize CSE + FeatureFlags option for run in background
```

### CODE REVIEW

**Comment by mboersma** [comment]
/lgtm

That's quite the one-liner!

File: pkg/api/vlabs/types.go [hunk 1]
```diff
@@ -42,6 +42,7 @@ type Properties struct {
 	CertificateProfile      *CertificateProfile      `json:"certificateProfile,omitempty"`
 	AADProfile              *AADProfile              `json:"aadProfile,omitempty"`
 	AzProfile               *AzProfile               `json:"azProfile,omitempty"`
+	Debug                   *Debug                   `json:"debug,omitempty"`
```
Comment (CecileRobertMichon) [inline pkg/api/vlabs/types.go]: thoughts on calling this "ExperimentalFeatures" or "FeatureFlags" instead? To me "debug" sounds like a bool that can turn on or off more debugging output

Comment (mboersma) [reply pkg/api/vlabs/types.go]: I made this same point offline. I thought of "Experimental" or "Incubator" (or "Danger") but I like @CecileRobertMichon's suggestions better.

Comment (jackfrancis) [reply pkg/api/vlabs/types.go]: How about TimeTrial?

Comment (tariq1890) [reply pkg/api/vlabs/types.go]: Or RunAsync

Comment (tariq1890) [reply pkg/api/vlabs/types.go]: IsRunInBackgroundEnabled?

Comment (CecileRobertMichon) [reply pkg/api/vlabs/types.go]: isn't this new field supposed to be a general "new features to include" struct? `TimeTrial` and `RunAsync` sound directly related to CSE async provisioning

Comment (CecileRobertMichon) [reply pkg/api/vlabs/types.go]: https://github.com/Azure/acs-engine/pull/4104/files/174b85b7870648b1abeaf550638844df360cbd94#diff-0226d5bd49c7195fda1784e5dd4274c9R58 is the specific one

Comment (tariq1890) [reply pkg/api/vlabs/types.go]: ok on a serious note, DebugSettings?

Comment (tariq1890) [reply pkg/api/vlabs/types.go]: CSEDebugSettings?

Comment (jackfrancis) [reply pkg/api/vlabs/types.go]: How about `Proto`? This is not meant to be just for "features". It is meant to represent  operational break glass scenarios, or "extra" tooling for testing/debugging.

Here are some examples:

- deliver load/performance workloads
- deliver long-running introspection (e.g., tcpdump)
- alternate/experimental implementations for A/B testing

File: pkg/acsengine/template_generator.go [hunk 2]
```diff
@@ -418,6 +418,12 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) templat
 		"IsHostedBootstrap": func() bool {
 			return false
 		},
+		"CSERunInBackground": func() bool {
+			if cs.Properties.FeatureFlags != nil {
+				return cs.Properties.FeatureFlags.EnableCSERunInBackground
+			}
+			return false
```
Comment (CecileRobertMichon) [inline pkg/acsengine/template_generator.go]: so for now this is disabled by default right?

**Comment by mboersma** [comment]
`FeatureFlags` works for me.

/lgtm

File: parts/k8s/kubernetesagentresourcesvmss.t [hunk 3]
```diff
@@ -155,7 +155,7 @@
...
-                  "commandToExecute": "[concat('for i in $(seq 1 1200); do if [ -f /opt/azure/containers/provision.sh ]; then break; fi; if [ $i -eq 1200 ]; then exit 100; else sleep 1; fi; done; ', variables('provisionScriptParametersCommon'),' GPU_NODE={{IsNSeriesSKU .}} /usr/bin/nohup /bin/bash -c \"/bin/bash /opt/azure/containers/provision.sh >> /var/log/azure/cluster-provision.log 2>&1\"')]"
+                  "commandToExecute": "[concat('retrycmd_if_failure() { r=$1; w=$2; t=$3; shift && shift && shift; for i in $(seq 1 $retries); do timeout $t ${@}; [ $? -eq 0  ] && break || if [ $i -eq $r ]; then return 1; else sleep $w; fi; done }; ERR_OUTBOUND_CONN_FAIL=50; retrycmd_if_failure 40 1 3 nc -vz k8s.gcr.io 443 || retrycmd_if_failure 40 1 3 nc -vz gcr.azk8s.cn 80 || exit $ERR_OUTBOUND_CONN_FAIL; for i in $(seq 1 1200); do if [ -f /opt/azure/containers/provision.sh ]; then break; fi; if [ $i -eq 1200 ]; then exit 100; else sleep 1; fi; done; ', variables('provisionScriptParametersCommon'),' GPU_NODE={{IsNSeriesSKU .}} /usr/bin/nohup /bin/bash -c \"/bin/bash /opt/azure/containers/provision.sh >> /var/log/azure/cluster-provision.log 2>&1{{if CSERunInBackground }} &{{end}}\"')]"
```
Comment (andyzhangx) [inline parts/k8s/kubernetesagentresourcesvmss.t]: ping (icmp protocol) on `gcr.azk8s.cn` is not available, shall I change to `bing.com` instead? I think ping `bing.com` would be more stable.

Comment (jackfrancis) [reply parts/k8s/kubernetesagentresourcesvmss.t]: it’s not ping, it’s a simple TCP 80 connection check

**Pull request merged**